cmd/kubectl-gadget/utils/trace.go

```diff
@@ -374,59 +377,101 @@ func SetTraceOperation(traceID string, operation string) error {
-		successNodeCount := 0
-		timeout := time.Since(start) > traceTimeout
-		nodeErrors := make(map[string]string)
-		nodeWarnings := make(map[string]string)
+	var watchOptions = metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", GLOBAL_TRACE_ID, traceID),
+	}
 
-		traces, err := getTraceListFromID(traceID)
-		if err != nil {
-			return nil, err
+	watcher, err := traceClient.GadgetV1alpha1().Traces("gadget").Watch(context.TODO(), watchOptions)
+	if err != nil {
+		return watch.NewEmptyWatch(), err
+	}
+
+	return watcher, nil
+}
+
+// waitForTraceState wait for traces whom ID is given as parameter to be in the
+// expected state.
+func waitForTraceState(traceID string, expectedState string) (gadgetv1alpha1.TraceList, error) {
+	var returnedTraces gadgetv1alpha1.TraceList
+
+	traces, err := getTraceListFromID(traceID)
+	if err != nil {
+		return gadgetv1alpha1.TraceList{}, err
+	}
+
+	watchedTracesNumber := 0
+	tracesNumber := len(traces.Items)
+
+	watcher, err := getTraceWatcher(traceID)
+	if err != nil {
+		return gadgetv1alpha1.TraceList{}, err
+	}
+
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
+	_, err = watchtools.UntilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error){
+		// Deal particularly with error.
+		if event.Type == watch.Error {
+			return false, err
 		}
 
-		for _, i := range traces.Items {
-			if i.Status.OperationError != "" {
-				nodeErrors[i.Spec.Node] = i.Status.OperationError
-			} else if i.Status.OperationWarning != "" {
-				nodeWarnings[i.Spec.Node] = i.Status.OperationWarning
-				// TODO(francis) This code will not work if the trace is already in the
-				// expected state, for example if we decide to generate it twice.
-				// We need to add a cookie (and check it) to be sure the trace is ready.
-			} else if i.Status.State == expectedState {
-				successNodeCount++
-			} else {
-				// Consider Trace as timed out if it neither moved the state forward
-				// nor notified of an error or warning within the time window.
-				if timeout {
-					nodeErrors[i.Spec.Node] = fmt.Sprintf("No results received from trace within %v", traceTimeout)
-					continue
-				}
+		// We are only interested in Added and Modified event, as we want
+		// Status.State value to change.
+		// More particularly, we monitor Added event for gadget like
+		// process-collector and Modified for gadget like seccompadvisor.
+		if event.Type != watch.Added && event.Type != watch.Modified {
+			return false, nil
+		}
 
-				time.Sleep(100 * time.Millisecond)
-				continue RetryLoop
-			}
+		trace := event.Object.(*gadgetv1alpha1.Trace)
+
+		if trace.Status.OperationWarning != "" {
+			fmt.Fprintf(os.Stderr, "%v\n", trace.Status.OperationWarning)
+
+			return false, nil
 		}
 
-		// Print errors even if some nodes succeeded.
-		defer printTraceFeedback(nodeErrors, len(traces.Items))
+		if trace.Status.OperationError != "" {
+			return false, errors.New(trace.Status.OperationError)
+		}
 
```

Comment:
I don't think this should return that soon. We still can get good information from other traces even if there is one on error.